### DISCUSSION TITLE
How do I warm up the cache properly?

### DISCUSSION THREAD

**Discussion by shizzard** [serenity-rs/serenity#2808] [Q&A] [answered]
TL;DR: I cannot find a proper way to get a message author display name. Permissions are `all`, bot permissions parameters are all set to be privileged.

The `msg.member(&ctx).await` is never `Ok(...)` (actually, it is always an `Err()` with reason `The required item is missing from the cache.`).
I know that discord only sends partial guild information when client is connected, that's why there is a [Request Guild Members](https://discord.com/developers/docs/topics/gateway-events#request-guild-members) event on a gateway. However, it seems that this event is never sent, and the members list is not populated.
For me it seems that I'm doing something wrong and missing some crucial step here.

**Comment by GnomedDev** [comment, marked as answer]
Message::member only returns MissingItem if the Message::guild_id is None, does that help you?

**Reply by shizzard** [reply]
Uh, yes, it seems correct, for some reason the `guild_id` in the message is `None`.
Short introduction: I'm building a simple command that collects last N messages from the channel and reports a summary with a call to OpenAI API.
The message list I'm getting in the prototype is the following:
```rust
ctx.http.get_messages(msg.channel_id, Some(MessagePagination::Before(msg.id)), Some(100)).await?;
```
These messages are in the guild channel, but the `guild_id` field is `None`.
Is that a flaw of discord API?..

NB: For some reason the `msg.message_reference.guild_id` field is `Some(...)`, but the `msg.referenced_message.guild_id` is `None`.
JFYI, a stripped data:
```rust
    Message {
        id: MessageId(...),
        channel_id: ChannelId(...),
        author: User {...},
        content: "...",
        ...
        message_reference: Some(
            MessageReference {
                message_id: Some(MessageId(...),),
                channel_id: ChannelId(...),
                guild_id: Some(GuildId(...),),
                fail_if_not_exists: None,
            },
        ),
        flags: Some(...),
        referenced_message: Some(
            Message {
                id: MessageId(...),
                channel_id: ChannelId(...),
                author: User {
                    ...
                    member: None,
                },
                ...
                guild_id: None,
                member: None,
            },
        ),
        ...
        guild_id: None,
        member: None,
    }
```

**Reply by jamesbt365** [reply]
The guild_id in message in none because it was not received over the gateway (the `MessageCreate` event), the docs on the `Message` struct outlines this.

**Reply by shizzard** [reply]
Yes, that's right.
The workaround for anyone facing the same problem in the future: use `ctx.http.get_member` method and pass the original message (from the `MessageCreate` event) `msg.guild_id.expect(...)` as the first parameter.
Thank you!
